recipes/LibriSpeech/ASR/CTC/train_from_scratch.py

```diff
@@ -0,0 +1,547 @@
+    def compute_forward(self, batch, stage):
+        """Forward computations from the waveform batches to the output probabilities."""
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+
+        # compute features
+        feats = self.hparams.compute_features(wavs)
+        current_epoch = self.hparams.epoch_counter.current
+        feats = self.modules.normalize(feats, wav_lens, epoch=current_epoch)
+
+        if stage == sb.Stage.TRAIN:
+            if hasattr(self.hparams, "augmentation"):
+                feats = self.hparams.augmentation(feats)
+
+        # forward modules
+        src = self.modules.CNN(feats)
+
+        enc_out, pred = self.modules.Transformer(
+            src, tgt=None, wav_len=wav_lens
+        )
+
+        # output layer for ctc log-probabilities
+        logits = self.modules.ctc_lin(enc_out)
+        p_ctc = self.hparams.log_softmax(logits)
+
+        return p_ctc, wav_lens
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC) given predictions and targets."""
+
+        p_ctc, wav_lens = predictions
+
+        ids = batch.id
+        tokens, tokens_lens = batch.tokens
+
+        loss = self.hparams.ctc_cost(p_ctc, tokens, wav_lens, tokens_lens)
+
+        if stage != sb.Stage.TRAIN:
+            # Decode token terms to words
+            sequence = sb.decoders.ctc_greedy_decode(
+                p_ctc, wav_lens, blank_id=self.hparams.blank_index
+            )
+            predicted_words = [
+                self.tokenizer.decode_ids(hyp).split(" ") for hyp in sequence
+            ]
+
+            # Convert indices to words
+            target_words = undo_padding(tokens, tokens_lens)
+            target_words = [
+                self.tokenizer.decode_ids(words).split(" ")
+                for words in target_words
+            ]
+
+            self.wer_metric.append(ids, predicted_words, target_words)
+            self.cer_metric.append(ids, predicted_words, target_words)
+
+        return loss
+
+    def on_evaluate_start(
+        self,
+        max_key=None,
+        min_key=None,
+        nb_avg_ckpt=1,
+        recoverable_name="model",
+    ):
+        """perform checkpoint averge if needed"""
+        super().on_evaluate_start()
+
+        ckpts = self.checkpointer.find_checkpoints(
+            max_key=max_key, min_key=min_key, max_num_checkpoints=nb_avg_ckpt,
+        )
+        ckpt = sb.utils.checkpoints.average_checkpoints(
+            ckpts, recoverable_name="model", device=self.device
+        )
+
+        self.hparams.model.load_state_dict(ckpt, strict=True)
+        self.hparams.model.eval()
+
+        msg = "Loaded an average of " + str(nb_avg_ckpt) + " checkpoints..."
+        logger.info(msg)
+
+    def evaluate(
+        self,
+        test_set,
+        max_key=None,
+        min_key=None,
+        progressbar=None,
+        test_loader_kwargs={},
+        nb_avg_ckpt=1,
+        recoverable_name="model",
+    ):
```

Comment:
Hello @shucongzhang, I have some hard time understanding the point of redefining these functions. Could you please elaborate a bit more? I'd like to move forward and update this PR with SB 1.0 but Idk if this is necessary. Thanks :) 
CC: @TParcollet maybe you do know why?

Reply:
Hello @Adel-Moumen , thank you so much for the review and really sorry for the late reply. I was not feeling well and then occupied by something else. I suppose the main reason is at line 538 
 `for nb_ckpt in [1, 5, 10]:`
we can have a quick and easy control about different number of models we want to do the model averaging. Thanks :)

Reply:
Hello @shucongzhang, hope you are getting better! It's fine, health comes first :)

Yes, at the end, I did understand why you modified it. I decided to remove it as I don't think it should be like that in SpeechBrain, but I do understand why you did it.